src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,903 @@
+  int loop_iter = num_images * 5;
+  size_t row_idx = 0;
+  int col_i = 0, col_s = 0, col_b = 0;
+  for(size_t i = 0; i < loop_iter; ++i)
+  {
+    if (i % 5 == 4)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(segmentations.colptr(col_s),\
+                                  col_size, row_size).t();
+      ++col_s;
+    }
+    else if (i % 5 == 3)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(boundaries.colptr(col_b),\
+                                  col_size, row_size).t();
+      ++col_b;
+    }
+    else
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(images.colptr(col_i),
+                                  col_size, row_size).t();
+      ++col_i;  
+    }
+    row_idx += row_size;
+  }
+  return input_data;
+} 
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+GetFeatureDimension()
+{
+  /*
+  shrink: amount to shrink channels
+  p_size: size of image patches
+  n_cell: number of self similarity cells
+  n_orient: number of orientations per gradient scale
+  */
+  arma::vec P(2);
+  int shrink, p_size, n_cell;
+  shrink = this->options["shrink"];
+  p_size = this->options["p_size"];
+  n_cell = this->options["n_cell"];
+
+  /*
+  n_color_ch: number of color channels
+  n_grad_ch: number of gradient channels
+  n_ch: total number of channels
+  */
+  int n_color_ch, n_grad_ch, n_ch;
+  if (this->options["rgbd"] == 0)
+    n_color_ch = 3;
+  else
+    n_color_ch = 4;
+
+  n_grad_ch = 2 * (1 + this->options["n_orient"]);
+
+  n_ch = n_color_ch + n_grad_ch;
+  P[0] = pow((p_size / shrink) , 2) * n_ch;
+  P[1] = pow(n_cell , 2) * (pow (n_cell, 2) - 1) / 2 * n_ch;
+  return P;
+}
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+dt_1d(arma::vec& f, int n)
+{
+  arma::vec d(n), v(n), z(n + 1);
+  int k = 0;
+  v[0] = 0.0;
+  z[0] = -INF;
+  z[1] = +INF;
+  for (size_t q = 1; q <= n - 1; ++q)
+  {
+    float s  = ( (f[q] + q * q)-( f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    while (s <= z[k])
+    {
+      --k;
+      s  = ( (f[q] + q * q) - (f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    }
+
+    k++;
+    v[k] = (double)q;
+    z[k] = s;
+    z[k+1] = +INF;
+  }
+
+  k = 0;
+  for (int q = 0; q <= n-1; q++)
+  {
+    while (z[k+1] < q)
+      k++;
+    d[q] = (q - v[k]) * (q - v[k]) + f[v[k]];
+  }
+  return d;
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+dt_2d(MatType& im)
+{
+  arma::vec f(std::max(im.n_rows, im.n_cols));
+  // transform along columns
+  for (size_t x = 0; x < im.n_cols; ++x)
+  {
+    f.subvec(0, im.n_rows - 1) = im.col(x);
+    arma::vec d = this->dt_1d(f, im.n_rows);
+    im.col(x) = d;
+  }
+
+  // transform along rows
+  for (int y = 0; y < im.n_rows; y++)
```

Comment:
I guess this line would emit warning(compare with unsigned and signed, change y to size_t or uword may solve this problem, I remember rcurtin prefer size_t since it is a standard way to store the size of array)